prow/bugzilla/client_test.go

```diff
@@ -930,3 +931,162 @@ func TestGetExternalBugPRsOnBug(t *testing.T) {
+			map[int]struct{}{1843407: exists, 1844101: exists},
+		},
+	}
+
+	testServer := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("X-BUGZILLA-API-KEY") != "api-key" {
+			t.Error("did not get api-key passed in X-BUGZILLA-API-KEY header")
+			http.Error(w, "403 Forbidden", http.StatusForbidden)
+			return
+		}
+		if r.URL.Query().Get("api_key") != "api-key" {
+			t.Error("did not get api-key passed in api_key query parameter")
+			http.Error(w, "403 Forbidden", http.StatusForbidden)
+			return
+		}
+		if r.Method != http.MethodGet {
+			t.Errorf("incorrect method to get a bug: %s", r.Method)
+			http.Error(w, "400 Bad Request", http.StatusBadRequest)
+			return
+		}
+		if !strings.HasPrefix(r.URL.Path, "/rest/bug/") {
+			t.Errorf("incorrect path to get a bug: %s", r.URL.Path)
+			http.Error(w, "400 Bad Request", http.StatusBadRequest)
+			return
+		}
+		if strings.HasPrefix(r.URL.Path, "/rest/bug/") {
+			id, err := strconv.Atoi(strings.TrimPrefix(r.URL.Path, "/rest/bug/"))
+			if err != nil {
+				t.Errorf("malformed bug id: %s", r.URL.Path)
+				http.Error(w, "400 Bad Request", http.StatusBadRequest)
+				return
+			}
+			if r.URL.Query().Get("include_fields") == "external_bugs" {
+				if val, ok := testcases[id]; ok {
+					w.Write([]byte(val.prDetails))
+				} else {
+					http.Error(w, "404 Not Found", http.StatusNotFound)
+				}
+			} else {
+				if val, ok := testcases[id]; ok {
+					w.Write([]byte(val.bugDetails))
+				} else {
+					http.Error(w, "404 Not Found", http.StatusNotFound)
+				}
+			}
+
+		}
+
+	}))
+	defer testServer.Close()
+	client := clientForUrl(testServer.URL)
+	for tc, tp := range testcases {
+		t.Run(strconv.Itoa(tc), func(t *testing.T) {
+			var parsedResponse struct {
+				Bugs []*Bug `json:"bugs,omitempty"`
+			}
+			if err := json.Unmarshal([]byte(tp.bugDetails), &parsedResponse); err != nil {
+				t.Fatal(err)
+			}
+			clones, err := client.GetAllClones(parsedResponse.Bugs[0])
+			if err != nil {
+				t.Errorf("Error occurred when none was expected: %v", err)
+			}
+			if len(tp.expectedClones) != len(clones) {
+				t.Errorf("Mismatch in number of clones - expected: %d, got %d", len(tp.expectedClones), len(clones))
+			}
+			for _, clone := range clones {
+				if _, ok := tp.expectedClones[clone.ID]; !ok {
+					t.Errorf("Unexpected clone found in list - expecting: %v, got: %d", tp.expectedClones, clone.ID)
+				}
+			}
+		})
+
+	}
+
+}
+
+func TestGetRootForClone(t *testing.T) {
+	bugMappings := map[int]string{
```

Comment:
These tests are pretty hard to read because they are based on embedded raw json and from looking at them its not immediately obvious which one covers which scenario like "direct parent", "root is a grandparent" etc.

A better way to do this would be to not test the whole handler but just the `getAllClones` function. The code is factored in a way that this is easely possible by passing in a fake/mock to `getAllClones`

Reply:
IMHO the same applies to the other testcase, it would be much easier to read if it wasn't based on embedded json

Reply:
I'm not sure I understand what you mean by this. As far as I understand I am testing only the GetAllClones function. Do you perhaps have an example for some other piece of code you are testing which would help me understand?
I can whittle down the size of the raw json if that would help, because not all the fields are necessary for my test.

Reply:
You are testing the `GetAllClones`, not the `getAllCloens`. The latter allows to inject a client for which you can use the `Fake` and on it you can define the individual bugs it will return. You can then only define the relevant fields there like `DependsOn` which should make the tests more readable